src/include/buffer.h

```diff
@@ -379,33 +381,305 @@ namespace buffer CEPH_BUFFER_API {
   };
 
 
+  struct ptr_hook {
+    mutable ptr_hook* next;
+  };
+
+  class ptr_node : public ptr_hook, public ptr {
+  public:
+    struct cloner {
+      ptr_node* operator()(const ptr_node& clone_this) {
+	return new ptr_node(clone_this);
+      }
+    };
+    struct disposer {
+      void operator()(ptr_node* const delete_this) {
+	if (!dispose_if_hypercombined(delete_this)) {
+	  delete delete_this;
+	}
+      }
+    };
+
+    ~ptr_node() = default;
+
+    static std::unique_ptr<ptr_node, disposer> create(raw* const r) {
+      return create_hypercombined(r);
+    }
+    static std::unique_ptr<ptr_node, disposer> create(const unsigned l) {
+      return create_hypercombined(buffer::create(l));
+    }
+    template <class... Args>
+    static std::unique_ptr<ptr_node, disposer> create(Args&&... args) {
+      return std::unique_ptr<ptr_node, disposer>(
+	new ptr_node(std::forward<Args>(args)...));
+    }
+
+  private:
+    template <class... Args>
+    ptr_node(Args&&... args) : ptr(std::forward<Args>(args)...) {
+    }
+    ptr_node(const ptr_node&) = default;
+
+    ptr& operator= (const ptr& p) = delete;
+    ptr& operator= (ptr&& p) noexcept = delete;
+    ptr_node& operator= (const ptr_node& p) = delete;
+    ptr_node& operator= (ptr_node&& p) noexcept = delete;
+
+    static bool dispose_if_hypercombined(ptr_node* delete_this);
+    static std::unique_ptr<ptr_node, disposer> create_hypercombined(raw* r);
+  };
   /*
    * list - the useful bit!
    */
 
   class CEPH_BUFFER_API list {
+  public:
+    // this the very low-level implementation of singly linked list
+    // ceph::buffer::list is built on. We don't use intrusive slist
+    // of Boost (or any other 3rd party) to save extra dependencies
+    // in our public headers.
+    class buffers_t {
+      // _root.next can be thought as _head
+      ptr_hook _root;
+      ptr_hook* _tail;
+      std::size_t _size;
+
+    public:
+      template <class T>
+      class buffers_iterator {
+	typename std::conditional<
+	  std::is_const<T>::value, const ptr_hook*, ptr_hook*>::type cur;
+	template <class U> friend class buffers_iterator;
+      public:
+	using value_type = T;
+	using reference = std::add_lvalue_reference_t<T>;
```

Comment:
`std::add_lvalue_reference_t` is introduced by C++14. and this breaks C++11 librados client.
```
2018-12-02T07:40:55.943 INFO:tasks.workunit.client.0.smithi168.stderr:In file included from /usr/include/rados/librados.hpp:11:0,
2018-12-02T07:40:55.943 INFO:tasks.workunit.client.0.smithi168.stderr:                 from /home/ubuntu/cephtest/mnt.0/client.0/tmp/rocksdb/include/rocksdb/utilities/env_librados.h:14,
2018-12-02T07:40:55.943 INFO:tasks.workunit.client.0.smithi168.stderr:                 from /home/ubuntu/cephtest/mnt.0/client.0/tmp/rocksdb/utilities/env_librados.cc:4:
2018-12-02T07:40:55.943 INFO:tasks.workunit.client.0.smithi168.stderr:/usr/include/rados/buffer.h:455:20: error: expected type-specifier
2018-12-02T07:40:55.943 INFO:tasks.workunit.client.0.smithi168.stderr:  using reference = std::add_lvalue_reference_t<T>;
2018-12-02T07:40:55.944 INFO:tasks.workunit.client.0.smithi168.stderr:                    ^
2018-12-02T07:40:55.944 INFO:tasks.workunit.client.0.smithi168.stderr:/usr/include/rados/buffer.h:456:18: error: expected type-specifier
2018-12-02T07:40:55.944 INFO:tasks.workunit.client.0.smithi168.stderr:  using pointer = std::add_pointer_t<T>;
2018-12-02T07:40:55.944 INFO:tasks.workunit.client.0.smithi168.stderr:                  ^
2018-12-02T07:40:55.946 INFO:tasks.workunit.client.0.smithi168.stderr:In file included from /usr/include/c++/4.8.2/bits/stl_algobase.h:65:0,
2018-12-02T07:40:55.947 INFO:tasks.workunit.client.0.smithi168.stderr:                 from /usr/include/c++/4.8.2/memory:62,
2018-12-02T07:40:55.947 INFO:tasks.workunit.client.0.smithi168.stderr:                 from /home/ubuntu/cephtest/mnt.0/client.0/tmp/rocksdb/include/rocksdb/utilities/env_librados.h:8,
2018-12-02T07:40:55.947 INFO:tasks.workunit.client.0.smithi168.stderr:                 from /home/ubuntu/cephtest/mnt.0/client.0/tmp/rocksdb/utilities/env_librados.cc:4:
2018-12-02T07:40:55.947 INFO:tasks.workunit.client.0.smithi168.stderr:/usr/include/c++/4.8.2/bits/stl_iterator_base_types.h: In instantiation of â€˜struct std::__iterator_traits<ceph::buffer::list::buffers_t::buffers_iterator<ceph::buffer::ptr_node>, true>â€™:
2018-12-02T07:40:55.947 INFO:tasks.workunit.client.0.smithi168.stderr:/usr/include/c++/4.8.2/bits/stl_iterator_base_types.h:159:12:   required from â€˜struct std::iterator_traits<ceph::buffer::list::buffers_t::buffers_iterator<ceph::buffer::ptr_node> >â€™
2018-12-02T07:40:55.947 INFO:tasks.workunit.client.0.smithi168.stderr:/usr/include/c++/4.8.2/bits/stl_iterator_base_funcs.h:184:5:   required by substitution of â€˜template<class _ForwardIterator> _ForwardIterator std::next(_ForwardIterator, typename std::iterator_traits<_Iter>::difference_type) [with _ForwardIterator = ceph::buffer::list::buffers_t::buffers_iterator<ceph::buffer::ptr_node>]â€™
2018-12-02T07:40:55.948 INFO:tasks.workunit.client.0.smithi168.stderr:/usr/include/rados/buffer.h:639:35:   required from here
2018-12-02T07:40:55.948 INFO:tasks.workunit.client.0.smithi168.stderr:/usr/include/c++/4.8.2/bits/stl_iterator_base_types.h:154:53: error: no type named â€˜pointerâ€™ in â€˜class ceph::buffer::list::buffers_t::buffers_iterator<ceph::buffer::ptr_node>â€™
2018-12-02T07:40:55.948 INFO:tasks.workunit.client.0.smithi168.stderr:       typedef typename _Iterator::pointer           pointer;
2018-12-02T07:40:55.948 INFO:tasks.workunit.client.0.smithi168.stderr:                                                     ^
2018-12-02T07:40:55.948 INFO:tasks.workunit.client.0.smithi168.stderr:/usr/include/c++/4.8.2/bits/stl_iterator_base_types.h:155:53: error: no type named â€˜referenceâ€™ in â€˜class ceph::buffer::list::buffers_t::buffers_iterator<ceph::buffer::ptr_node>â€™
2018-12-02T07:40:55.948 INFO:tasks.workunit.client.0.smithi168.stderr:       typedef typename _Iterator::reference         reference;
2018-12-02T07:40:55.948 INFO:tasks.workunit.client.0.smithi168.stderr:                                                     ^
2018-12-02T07:40:56.209 INFO:tasks.workunit.client.0.smithi168.stderr:make[3]: *** [CMakeFiles/rocksdb-shared.dir/utilities/env_librados.cc.o] Error 1
```

see http://pulpito.ceph.com/kchai-2018-12-02_07:20:25-rados-wip-kefu2-testing-2018-12-02-0912-distro-basic-smithi/3302735/

Reply:
Thanks for pointing out the issues and apologize for unnecessary burden. I fixed the C++11 compliance and a leak in `::rebuild_aligned_size_and_memory` I was able to replicate locally. Hopefully it was responsible also for the failures at Teuthology.

```cpp
int FileJournal::prepare_entry(vector<ObjectStore::Transaction>& tls, bufferlist* tbl) {
  // ...
  bufferlist ebl;
  // header
  ebl.append((const char*)&h, sizeof(h));

  // ...

  if (directio)
    ebl.rebuild_aligned(CEPH_DIRECTIO_ALIGNMENT);
  // ...
}
```

```cpp
  bool buffer::list::rebuild_aligned(unsigned align)
  {
    return rebuild_aligned_size_and_memory(align, align);
  }
```

Reply:
Just scheduled a [tiny run](http://pulpito.front.sepia.ceph.com/rzarzynski-2018-12-03_10:41:05-rados-wip-bl-hypercombined-distro-basic-smithi/) to preliminarily verify the fix.

Reply:
The single job looks green now.